lib/OpenStackExtract/MiqOpenStackVm/MiqOpenStackInstance.rb

```diff
@@ -0,0 +1,188 @@
+require 'util/miq_tempfile'
+require_relative '../../MiqVm/MiqVm'
+
+#
+# TODO: Create common base class for MiqOpenStackInstance and MiqOpenStackImage
+# and factor out common code. Also, refactor MiqVm so it can be a proper super class.
+#
+
+class MiqOpenStackInstance
+  SUPPORTED_METHODS = [:vmRootTrees, :extract, :diskInitErrors]
+
+  def initialize(instance_id, args)
+    @instance_id      = instance_id
+    @args             = args
+    @openstack_handle = args[:openstack_handle]
+    raise ArgumentError, "#{self.class.name}: required arg openstack_handle missing"  unless @openstack_handle
+  end
+
+  def compute_service
+    @compute_service ||= @openstack_handle.compute_service
+  end
+
+  def image_service
+    @image_service ||= @openstack_handle.detect_image_service
+  end
+
+  def instance
+    @instance ||= compute_service.servers.get(@instance_id)
+  end
+
+  def snapshot_metadata
+    @snapshot_metadata ||= instance.metadata.get(:miq_snapshot)
+  end
+
+  def snapshot_image_id
+    @snapshot_image_id ||= snapshot_metadata && snapshot_metadata.value
+  end
+
+  def unmount
+    return unless @miq_vm
+    @miq_vm.unmount
+    @temp_image_file.unlink
+  end
+
+  def create_evm_snapshot(options = {})
+    log_prefix = "MIQ(#{self.class.name}##{__method__}) instance_id=[#{@instance_id}]"
+
+    miq_snapshot = nil
+    if snapshot_image_id
+      $log.debug "#{log_prefix}: Found pointer to existing snapshot: #{snapshot_image_id}"
+      miq_snapshot = begin
+        image_service.images.get(snapshot_image_id)
+      rescue => err
+        $log.debug "#{log_prefix}: #{err}"
+        $log.debug err.backtrace.join("\n")
```

Comment:
Should this be debug or error?  Maube the first line is error and the backtrace should be debug.

Additionally, should we start protecting ourselves in this code (and lib in general) from $log?  A simple way to do it would be to make a `def logger` instance method that returns `@logger ||= $log` and perhaps a `def logger=` setter to set `@logger`.  Then the code is in place to make the logger a settable thing in the future, and to disconnect the lib from from the knowledge of what the logger object is.